### PR TITLE
Hk cyclomatic complexity

### PR DESCRIPTION
**Pull request by hanneskaeufler**
Adresses #91. Just fyi that I tried this. Uses the same logic as rubocop does. I might go ahead and extract this CountingVisitor into a separate file and add more test coverage. I kind of works 🤷🏼‍♂️ Was easy, mad props on the architecture!

### CODE REVIEW

File: src/ameba/rule/metric/cyclomatic_complexity.cr [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+      def count
+        @scope.accept(self)
+        @complexity
+      end
+
+      # Uses the same logic than rubocop. See
+      # https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/metrics/cyclomatic_complexity.rb#L21
+      {% for node in [:if, :while, :until, :for, :rescue, :when, :or, :and] %}
+      def visit(node : Crystal::{{ node.id.capitalize }})
```
Comment (Sija) [inline src/ameba/rule/metric/cyclomatic_complexity.cr]: Indent plz.

File: src/ameba/rule/metric/cyclomatic_complexity.cr [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      end
+
+      def count
+        @scope.accept(self)
+        @complexity
+      end
+
+      # Uses the same logic than rubocop. See
+      # https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/metrics/cyclomatic_complexity.rb#L21
+      {% for node in [:if, :while, :until, :for, :rescue, :when, :or, :and] %}
```
Comment (Sija) [inline src/ameba/rule/metric/cyclomatic_complexity.cr]:
```suggestion
      {% for node in %i(if while until for rescue when or and) %}
```

File: src/ameba/rule/metrics/cyclomatic_complexity.cr [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    def test(source)
+      AST::NodeVisitor.new self, source
+    end
+
+    def test(source, node : Crystal::Def)
+      complexity = AST::CountingVisitor.new(node).count
+
+      if complexity > max_complexity
+        issue_for(node, MSG % [complexity, max_complexity])
```
Comment (Sija) [inline src/ameba/rule/metrics/cyclomatic_complexity.cr]:
```suggestion
        issue_for node, MSG % {complexity, max_complexity}
```

Comment (Sija) [reply src/ameba/rule/metrics/cyclomatic_complexity.cr]: `Tuple` is stack based, whereas `Array` is allocated on the heap increasing GC pressure.

Comment (hanneskaeufler) [reply src/ameba/rule/metrics/cyclomatic_complexity.cr]: Thanks!

**Comment by hanneskaeufler** [comment]
I think this is now ready for review, thanks for the comments!

**Comment by Sija** [comment]
`crystal tool format` :)

**Comment by hanneskaeufler** [comment]
Yes looks pretty horrible, but oh well 🤷🏼‍♂️

File: src/ameba/rule/metrics/cyclomatic_complexity.cr [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  #
+  # ```
+  # Metrics/CyclomaticComplexity:
+  #   Enabled: true
+  #   MaxComplexity: 10
+  # ```
+  #
+  struct CyclomaticComplexity < Base
+    properties do
+      enabled false
```
Comment (veelenga) [inline src/ameba/rule/metrics/cyclomatic_complexity.cr]: Let's enable that by default. Just remove this line, please.

**Comment by veelenga** [comment]
FYI one of the most complex methods in crystal repo (level of complexity: 513)

https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/syntax/lexer.cr#L90-L1229

**Comment by Sija** [comment]
@veelenga whooaaa, I thought that scrolling will never end...

File: src/ameba/rule/metrics/cyclomatic_complexity.cr (hunk 4 above)
Comment (hanneskaeufler) [reply src/ameba/rule/metrics/cyclomatic_complexity.cr]: Sure, happy to ✌️

**Comment by hanneskaeufler** [comment]
> FYI one of the most complex methods in crystal repo (level of complexity: 513)
> 
> https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/syntax/lexer.cr#L90-L1229

Haha that's a good one 🤣

Btw 10 as the threshold was just a wild guess. [Rubocop has is set to 6](https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml#L1677), [swiftlint warns at 10, errors at 20](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/Metrics/CyclomaticComplexityRule.swift#L5). [Some python tool says 10 is still a B](https://radon.readthedocs.io/en/latest/api.html#radon.complexity.cc_rank). I'd be fine with 10 as a default!

**Comment by veelenga** [comment]
Thank you guys. Good work

**Pull request merged**